### PR TITLE
Fix record discard/delete & subscribe race

### PR DESCRIPTION
**Pull request by ronag**
This is my try at fixing the discard/delete and subscribe data-races. https://github.com/deepstreamIO/deepstream.io-client-js/issues/204

The basic idea is that the record can revive itself AFTER successful discard/delete, using the pending queue for the time between re-subscription and revival.

This ensures that there CAN BE ONLY ONE active record instance for the client avoiding the:

```
UNSOLICITED_MESSAGE
UNSOLICITED_MESSAGE
ACK_TIMEOUT
RESPONSE_TIMEOUT
MULTIPLE_SUBSCRIBTIONS
```

error cycle

### CODE REVIEW

File: src/record/record-handler.js [hunk 1]
```diff
@@ -185,26 +183,6 @@ RecordHandler.prototype._$handle = function( message ) {
...
-			) {
-			this._destroyEventEmitter.emit( 'destroy_ack_' + name, message );
-
-			if( message.data[ 0 ] === C.ACTIONS.DELETE && this._records[ name ] ) {
-				this._records[ name ]._$onMessage( message );
-			}
-
-			return;
-		}
-
```
Comment (ronag) [inline src/record/record-handler.js]: Parts of this is probably still needed. Unclear.

File: src/record/record.js [hunk 2]
```diff
@@ -237,9 +240,10 @@ Record.prototype.discard = function() {
 	this.whenReady( function() {
 		this.usages--;
 		if( this.usages <= 0 ) {
-				this.emit( 'destroyPending' );
-				this._discardTimeout = setTimeout( this._onTimeout.bind( this, C.EVENT.ACK_TIMEOUT ), this._options.subscriptionTimeout );
-				this._connection.sendMsg( C.TOPIC.RECORD, C.ACTIONS.UNSUBSCRIBE, [ this.name ] );
+			this.isReady = false;
+			this._discardTimeout = setTimeout( this._onTimeout.bind( this, C.EVENT.ACK_TIMEOUT ), this._options.subscriptionTimeout );
+			this._connection.sendMsg( C.TOPIC.RECORD, C.ACTIONS.UNSUBSCRIBE, [ this.name ] );
+	  	this._resubscribeNotifier.destroy();
```
Comment (ronag) [inline src/record/record.js]: Not part of the original implementation. But I believe the resubscribeNotifier should be destroyed as soon as the `UNSUBSCRIBE`has been sent.

File: src/record/record.js [hunk 3]
```diff
@@ -628,14 +633,18 @@ Record.prototype._onTimeout = function( timeoutType ) {
...
- 	this._clearTimeouts();
- 	this._eventEmitter.off();
- 	this._resubscribeNotifier.destroy();
- 	this.isDestroyed = true;
- 	this.isReady = false;
- 	this._client = null;
-	this._eventEmitter = null;
-	this._connection = null;
+	if (this.usages > 0) {
+		this._init()
```
Comment (ronag) [inline src/record/record.js]: This is the magic. We allow a record to be revive AFTER it's life cycle has finished.

File: src/record/record.js [hunk 4]
```diff
@@ -255,9 +259,10 @@ Record.prototype.delete = function() {
 		return;
 	}
 	this.whenReady( function() {
-		this.emit( 'destroyPending' );
+		this.isReady = false;
```
Comment (ronag) [inline src/record/record.js]: Setting isReady to false allows the record to be revived with pending operations.

File: src/record/record.js [hunk 5]
```diff
@@ -237,9 +240,10 @@ Record.prototype.discard = function() {
 	this.whenReady( function() {
 		this.usages--;
 		if( this.usages <= 0 ) {
-				this.emit( 'destroyPending' );
-				this._discardTimeout = setTimeout( this._onTimeout.bind( this, C.EVENT.ACK_TIMEOUT ), this._options.subscriptionTimeout );
-				this._connection.sendMsg( C.TOPIC.RECORD, C.ACTIONS.UNSUBSCRIBE, [ this.name ] );
+			this.isReady = false;
```
Comment (ronag) [inline src/record/record.js]: Setting isReady to false allows the record to be revived with pending operations.

**Comment by ronag** [comment]
travis fails since the test don't allow re-using/reviving record instances.

**Comment by ronag** [comment]
This also resolve the issue of spamming subscribe/unsubscribe as there as a natural throttling.

File: src/record/record.js [hunk 6]
```diff
@@ -628,14 +628,19 @@ Record.prototype._onTimeout = function( timeoutType ) {
...
- 	this.isDestroyed = true;
- 	this.isReady = false;
- 	this._client = null;
-	this._eventEmitter = null;
-	this._connection = null;
+	if (this.usages > 0) {
+		this._sendRead();
+	}
+	else {
+		this._clearTimeouts();
```
Comment (ronag) [inline src/record/record.js]: Clear timeout beredda to be moved to top and always be invoked.

**Comment by ronag** [comment]
@yasserf: tests pass

**Comment by ronag** [comment]
@yasserf: This PR has a race when disconnecting the client. I'm currently in a state where I have no working solution to this issue.

**Comment by ronag** [comment]
I'm closing this as it's broken.

**Comment by ronag** [comment]
Fixed issue.

File: src/utils/resubscribe-notifier.js [hunk 7]
```diff
@@ -40,14 +40,14 @@ ResubscribeNotifier.prototype.destroy = function() {
...
-		
-	if( state === C.CONNECTION_STATE.RECONNECTING && this._isReconnecting === false ) {
-		this._isReconnecting = true;
-	}
+
 	if( state === C.CONNECTION_STATE.OPEN && this._isReconnecting === true ) {
 		this._isReconnecting = false;
 		this._resubscribe();
 	}
+	else if ( this._isReconnecting === false ) {
```
Comment (ronag) [inline src/utils/resubscribe-notifier.js]: We always want to reconnect when going from non OPEN to OPEN.

**Comment by ronag** [comment]
#232

File: src/record/record.js [hunk 8]
```diff
@@ -42,11 +42,7 @@ var Record = function( name, recordOptions, connection, options, client ) {
 	if( options.mergeStrategy ) {
 		this.setMergeStrategy( options.mergeStrategy );
 	}
-
-	this._resubscribeNotifier = new ResubscribeNotifier( this._client, this._sendRead.bind( this ) );
-	this._readAckTimeout = setTimeout( this._onTimeout.bind( this, C.EVENT.ACK_TIMEOUT ), this._options.recordReadAckTimeout );
-	this._readTimeout = setTimeout( this._onTimeout.bind( this, C.EVENT.RESPONSE_TIMEOUT ), this._options.recordReadTimeout );
-	this._sendRead();
+	this._resubscribeNotifier = new ResubscribeNotifier( this._client, this._sendRead.bind( this ), this._clearTimeouts.bind( this ), true );
```
Comment (ronag) [inline src/record/record.js]: Simply doing a clearTimeouts on disconnect is not good enough...

**Comment by ronag** [comment]
I don't think this PR is good enough. But the basic ideas are there. There are some edge cases that I don't think are properly handled.

**Comment by yasserf** [comment]
Yeah, we need to look into the discard/delete async case in more detail

**Comment by ronag** [comment]
This needs to be solved in another way. We've resolved it locally in our fork in a different way.

**Pull request closed without merging**